gsa/src/web/components/form/useForm.js

```diff
@@ -0,0 +1,109 @@
+
+  const words = parsedString.split(' ');
+  let errorString = '';
+
+  words.forEach(word => (errorString += capitalizeFirstLetter(word) + ' '));
+  return errorString.trim(); // parse "name" to something human readable. Accepts underscores and camelcase. "credential_id" > "Credential Id", etc.
+};
+
+export const syncVariables = (values, formState, dependencies = {}) => {
+  Object.keys(formState).forEach(key => (values[key] = formState[key]));
+  Object.keys(dependencies).forEach(key => (values[key] = dependencies[key]));
+}; // sync all form states with the values known to SaveDialog
+
+const useForm = (stateSchema, validationRules, callback, extras = {}) => {
+  const [formState, setFormState] = useState(stateSchema);
+  const [dependencies, setDependencies] = useState(extras);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [shouldWarn, setWarn] = useState(false); // shouldWarn is false when first rendered. Only when calling handleSubmit for the first time will this be set to true.
+
+  const [formStatus, setFormStatus] = useState(stateSchema); // use the same shape as stateschema
+
+  const handleValueChange = (value, name) => {
+    setFormState(prevValues => ({...prevValues, [name]: value}));
+  };
+
+  const handleDependencyChange = (value, name) => {
+    setDependencies(prevValues => ({...prevValues, [name]: value}));
+  };
+
+  // eslint-disable-next-line no-shadow
+  const validate = (value, name, dependencies) => {
+    setFormStatus(validity => ({
+      ...validity,
+      [name]: validationRules[name](value, dependencies),
+    }));
+  };
+
+  useEffect(() => {
+    Object.keys(formState).forEach(key => {
+      validate(formState[key], key, dependencies);
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [formState, dependencies]);
+
+  const handleSubmit = vals => {
+    setWarn(true);
+    const errorVars = [];
+    Object.keys(formState).forEach(key => {
+      !formStatus[key].validity && errorVars.push(parseAlias(key)); // Collects form fields with errors
+    });
+
+    const hasErrorInState = Object.keys(formState).some(key => {
+      return formStatus[key].validity === false;
+    }, []); // checks if any field is invalid
+
+    if (hasErrorInState) {
+      setErrorMessage('Error in fields: ' + errorVars.join(', ').toString()); // Outputs an error string containing all errors, or does nothing at all except show error bubbles. This is good for multi step dialogs.
+    } else {
+      // eslint-disable-next-line callback-return
+      return callback(vals); // if nothing is wrong, call onSave
+    }
+  };
+
+  return {
+    dependencies,
+    errorMessage,
+    shouldWarn,
+    formStatus,
+    formState,
+    handleDependencyChange,
+    handleSubmit,
+    handleValueChange,
+  };
+};
+
+// Some common validation rules
+export const testNonEmptyString = string => string.length > 0 && string !== ' ';
```

Comment:
This will invalidate `' '` but not `'  '` or `'     '`

Reply:
Hm. maybe we should use a trim() function on the string and see if it's still length zero. And for that matter we should also invalidate stuff like ' hi'. Maybe.

Reply:
I would just trim ' hi' to 'hi' without even asking the user or something